bundles/org.openhab.binding.openthermgateway/src/main/java/org/openhab/binding/openthermgateway/internal/OpenThermGatewaySocketConnector.java

```diff
@@ -0,0 +1,220 @@
+
+    public OpenThermGatewaySocketConnector(OpenThermGatewayCallback callback, String ipaddress, int port) {
+        this.callback = callback;
+        this.ipaddress = ipaddress;
+        this.port = port;
+    }
+
+    @Override
+    public void run() {
+        stopping = false;
+        connected = false;
+
+        logger.debug("Connecting OpenThermGatewaySocketConnector to {}:{}", this.ipaddress, this.port);
+
+        callback.connecting();
+
+        try (Socket socket = new Socket()) {
+            socket.connect(new InetSocketAddress(this.ipaddress, this.port), COMMAND_TIMEOUT_MILLISECONDS);
+            socket.setSoTimeout(COMMAND_TIMEOUT_MILLISECONDS);
+
+            connected = true;
+
+            callback.connected();
+
+            logger.debug("OpenThermGatewaySocketConnector connected");
+
+            try (BufferedReader reader = new BufferedReader(new InputStreamReader(socket.getInputStream()));
+                    PrintWriter wrt = new PrintWriter(socket.getOutputStream(), true)) {
+                // Make writer accessible on class level
+                writer = wrt;
+
+                sendCommand(GatewayCommand.parse(GatewayCommandCode.PrintReport, "A"));
+                // Set the OTGW to report every message it receives and transmits
+                sendCommand(GatewayCommand.parse(GatewayCommandCode.PrintSummary, "0"));
+
+                while (!stopping && !Thread.currentThread().isInterrupted()) {
+                    @Nullable
+                    String message = reader.readLine();
+
+                    if (message != null) {
+                        handleMessage(message);
+                    } else {
+                        logger.debug("Connection closed by OpenTherm Gateway");
+                        break;
+                    }
+                }
+
+                logger.debug("Stopping OpenThermGatewaySocketConnector");
+            } finally {
+                connected = false;
+
+                logger.debug("OpenThermGatewaySocketConnector disconnected");
+                callback.disconnected();
+            }
+        } catch (IOException ex) {
+            logger.warn("Unable to connect to the OpenTherm Gateway.", ex);
+        }
+    }
+
+    @Override
+    public void stop() {
+        logger.debug("Stopping OpenThermGatewaySocketConnector");
+        stopping = true;
+    }
+
+    @Override
+    public boolean isConnected() {
+        return connected;
+    }
+
+    @Override
+    public void sendCommand(GatewayCommand command) {
+        @Nullable
+        PrintWriter wrtr = writer;
+
+        String msg = command.toFullString();
+
+        pendingCommands.put(command.getCode(),
+                new AbstractMap.SimpleImmutableEntry<>(System.currentTimeMillis(), command));
+
+        if (connected) {
+            logger.debug("Sending message: {}", msg);
+            if (wrtr != null) {
+                wrtr.printf("%s\r\n", msg);
```

Comment:
The writer is synchronized under the hood so you can't really split this up into two method calls. That said, you don't need to use `printf` for something this simple.
```suggestion
                wrtr.print(msg+"\r\n");
```

Reply:
Modified.. though I am not sure I understand your comment about splitting things up into two method calls.

Reply:
Sorry about that, I was half mutter to myself while writing the suggested change.
My original suggestion that I changed before posting would be something like this:
```
wrtr.print(msg);
wrtr.print("\r\n");
```
But I quickly realized that such a suggestion wouldn't be best since you could run into issues if that code was called by multiple threads at once. The `PrintWriter` will synchronize all method calls but that won't help you if you spread out our message across multiple method calls.

That is why I instead suggested doing the string concatenation before calling `print`.

It would have been better if I hadn't mentioned multiple method calls since it wasn't relevant to the suggestion so sorry for confusing you.